keps/sig-network/3015-prefer-same-node/README.md

```diff
@@ -0,0 +1,390 @@
+- [ ] (R) KEP approvers have approved the KEP status as `implementable`
+- [ ] (R) Design details are appropriately documented
+- [ ] (R) Test plan is in place, giving consideration to SIG Architecture and SIG Testing input (including test refactors)
+  - [ ] e2e Tests for all Beta API Operations (endpoints)
+  - [ ] (R) Ensure GA e2e tests meet requirements for [Conformance Tests](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/conformance-tests.md) 
+  - [ ] (R) Minimum Two Week Window for GA e2e tests to prove flake free
+- [ ] (R) Graduation criteria is in place
+  - [ ] (R) [all GA Endpoints](https://github.com/kubernetes/community/pull/1806) must be hit by [Conformance Tests](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/conformance-tests.md) 
+- [ ] (R) Production readiness review completed
+- [ ] (R) Production readiness review approved
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+<!--
+**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
+-->
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+This KEP extends KEP-4444 `TrafficDistribution` with a new value,
+`PreferSameNode`, indicating traffic for a service should
+preferentially be routed to endpoints on the same node as the client.
+
+(This is the third attempt at this feature, which was previously
+suggested as [`internalTrafficPolicy: PreferLocal`] and [Node-level
+topology].)
+
+[`internalTrafficPolicy: PreferLocal`]: https://github.com/kubernetes/enhancements/pull/3016
+[Node-level topology]: https://github.com/kubernetes/enhancements/pull/3293
+
+## Motivation
+
+### Goals
+
+- Allow configuring a service so that connections will be delivered to
+  a local endpoint when possible, and a remote endpoint if not.
+
+### Non-Goals
+
+N/A
+
+## Proposal
+
+### User Stories
+
+#### DNS
+
+As a cluster administrator, I plan to run a DNS pod on each node, and
+would like DNS requests from other pods to always go to the local DNS
+pod, for efficiency. However, if no local DNS pod is available, DNS
+should just go to a remote pod instead so it keeps working. There
+should never be enough DNS traffic to overload any one endpoint, so
+it's safe to use a TrafficDistribution mode that doesn't worry about
+endpoint overload.
+
+### Risks and Mitigations
+
+This is similar to the existing `PreferClose` mode (possibly to be
```

Comment:
I think this should be a part of this KEP.  Add new name, deprecate (but retain) the `PreferClose` alias.

Reply:
This KEP adds a new field to `discoveryv1.EndpointHints` so it needs 3 releases in alpha to deal with potential apiserver skew, right? So I think it would be better to do `PreferSameZone` as an update to KEP-4444 rather than tying it to this.